### PR TITLE
python-common: Add drive selection

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Replaces #28896

**TODO**:

* [ ] ~~Depends on #30662~~ Actually I'd say it doesn't depend on 30662 as long as both PRs push teh same inventory.py
* [x] Depends on #31071 
* [ ] Documentation
* [x] add tests
* [ ] squash commits. **Please don't review individual commits yet**

Rational is that this can then be used by:

* mgr/rook to properly support drive groups
* mgr/ansible to properly support drive groups
* mgr/orchestrator_cli to preview OSD creation
* mgr/dashboard to preview OSD creation
* ceph-volume can then add native support for drive goups

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/python-common/ceph/deployment/inventory.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+from functools import total_ordering
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/inventory.py]: This file originates from #30662 and will not be part of this PR

**Comment by jan--f** [comment]
What is the purpose of having a drivegroups plugin in ceph-volume?

**Comment by sebastian-philipp** [comment]
> What is the purpose of having a drivegroups plugin in ceph-volume?

There are still some left-overs in ceph-volume. I'm going to remove all of them from this PR.

File: doc/ceph-volume/drivegroups/README.rst [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+If for whatever reason we can't rely on the rotational flag, we may just use any other property.
+
+
+.. code-block:: yaml
+
+        default:
+            data_devices:
+              size: 2TB # high:low notation is also implemented (:N, N:, MIN:MAX)
+            db_devices:
+              size: 512MB # MB or M is fine
```
Comment (krig) [inline doc/ceph-volume/drivegroups/README.rst]: Should be 512GB

Comment (sebastian-philipp) [reply doc/ceph-volume/drivegroups/README.rst]: thanks! please don't review individual commits at this point

**Comment by sebastian-philipp** [comment]
cc @jschmid1 : do you want to review this one?

**Comment by sebastian-philipp** [comment]
@bk201 

> Maybe a bit off-topic. If we have the following hardware setup:
> 
> ```
> two nodes and their free devices:
> node1: 2 HDDs, 1 SSD
> node2: 2 HDDs, 0 SSD
> ```
> 
> What happen if we submit the following DriveGroupSpec to orchestrator backend like DeepSea and the dg is passed through to ceph-volume with this PR?
> 
> ```
> drive_group_default_name:
>   target: *
>   data_devices:
>     rotational: 1
>   db_devices:
>     rotational: 0
> ```
> 
> node2 has no free SSD to be allocated as db device, should the operation fail? Or the operation still succeed with node1 has 1 SSD shared but node2 has none?
> 
> Dashboard needs to allow user previewing the result of DriveGroup submission. If the operation will fail, the logic in Dashboard should not allow user click the submit button at all. But if the operation is succeeded with best effort, Dashboard needs to provide correct preview for every host.

At this level of this PR, both calls will succeed independently. So yeah, db_devices will be an empty list.

**Comment by sebastian-philipp** [comment]
QA passed: http://pulpito.ceph.com/swagner-2019-11-06_09:22:28-rados:mgr-wip-swagner-testing-distro-basic-smithi/

File: src/python-common/ceph/deployment/drive_selection/matchers.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        :param dict disk: A disk representation (see base for examples)
+        :return: True/False if the match succeeded
+        :rtype: bool
+        """
+        if not disk:
+            return False
+        disk_value = self._get_disk_key(disk)
+        if str(self.value) in disk_value:
+            return True
+        return False
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/drive_selection/matchers.py]: return str(self.value) in disk_value

File: src/python-common/ceph/deployment/drive_selection/matchers.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+
+        A rather dumb matcher that just accepts all disks
+        (regardless of the value)
+        :param dict disk: A disk representation (see base for examples)
+        :return: always True
+        :rtype: bool
+        """
+        if not disk:
+            return False
+        return True
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/drive_selection/matchers.py]: return bool(disk)

File: src/python-common/ceph/deployment/drive_group.py [hunk 5]
```diff
@@ -123,29 +190,70 @@ def __init__(self,
         self.osd_id_claims = osd_id_claims
 
     @classmethod
-    def from_json(self, json_drive_group):
+    def from_json(cls, json_drive_group):
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/drive_group.py]: I would include the filters and size validation in the validation method

File: src/python-common/ceph/deployment/drive_group.py [hunk 6]
```diff
@@ -123,29 +190,70 @@ def __init__(self,
...
+                    json_drive_group.items()}
+            return DriveGroupSpec(**args)
+        except (KeyError, TypeError) as e:
+            raise DriveGroupValidationError(str(e))
 
     def hosts(self, all_hosts):
+        # type: (List[str]) -> List[str]
         return fnmatch.filter(all_hosts, self.host_pattern)
 
     def validate(self, all_hosts):
```
Comment (jmolmo) [inline src/python-common/ceph/deployment/drive_group.py]: I would call this method when a new DriveGroupSpec object is created (via init or from_json) .

**Comment by sebastian-philipp** [comment]
@jmolmo mind, if we merge this one to unblock @bk201 ?

File: src/python-common/ceph/deployment/drive_selection/matchers.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/drive_selection/matchers.py]: agree

File: src/python-common/ceph/deployment/drive_group.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/drive_group.py]: agree

File: src/python-common/ceph/deployment/drive_group.py (hunk 6 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/drive_group.py]: agree

**Comment by jmolmo** [comment]
> @jmolmo You have valid points, but none of them are critical. Mind, if we merge this one to unblock @bk201 ?

Of course. Just minor caveats. Go ahead. LGTM.

File: src/python-common/ceph/tests/test_disk_selector.py [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+        ("G", 'GB'),
+        ("GB", 'GB'),
+        ("TB", 'TB'),
+        ("T", 'TB'),
+        ("MB", 'MB'),
+        ("M", 'MB'),
+    ])
+    def test_normalize_suffix(self, test_input, expected):
+
+        assert drive_selection.SizeMatcher('10G', 'size')._normalize_suffix(test_input) == expected
```
Comment (bk201) [inline src/python-common/ceph/tests/test_disk_selector.py]: Suppose to be `('size', '10G')`.

File: src/python-common/ceph/tests/test_disk_selector.py [hunk 8]
```diff
@@ -0,0 +1,760 @@
...
+
+    def test_normalize_suffix_raises(self):
+
+        with pytest.raises(ValueError):
+            drive_selection.SizeMatcher('10P', 'size')._normalize_suffix("P")
```
Comment (bk201) [inline src/python-common/ceph/tests/test_disk_selector.py]: Suppose to be ('size', '10P').

File: src/python-common/ceph/tests/test_disk_selector.py [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+    def test_compare_false(self):
+        disk_dict = Device(path='/dev/vdb', sys_api=dict(model='nothing_matching'))
+        matcher = drive_selection.SubstringMatcher('model', 'samsung')
+        ret = matcher.compare(disk_dict)
+        assert ret is False
+
+
+class TestEqualityMatcher(object):
+    def test_compare(self):
+        disk_dict = Device(path='/dev/vdb', sys_api=dict(rotates='1'))
```
Comment (bk201) [inline src/python-common/ceph/tests/test_disk_selector.py]: The attribute is renamed to `rotational`.

File: src/python-common/ceph/tests/test_disk_selector.py [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+        matcher = drive_selection.EqualityMatcher('rotates', '1')
+        ret = matcher.compare(disk_dict)
+        assert ret is True
+
+    def test_compare_false(self):
+        disk_dict = Device(path='/dev/vdb', sys_api=dict(rotates='1'))
```
Comment (bk201) [inline src/python-common/ceph/tests/test_disk_selector.py]: The attribute is renamed to `rotational`.

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+                 ):
+        self.disks = disks.copy()
+        self.spec = spec
+
+        self._data = self.assign_devices(self.spec.data_devices)
+        self._wal = self.assign_devices(self.spec.wal_devices)
+        self._db = self.assign_devices(self.spec.db_devices)
+        self._jornal = self.assign_devices(self.spec.journal_devices)
+
+    def data_devices(self):
```
Comment (votdev) [inline src/python-common/ceph/deployment/drive_selection/selector.py]: Wouldn't it make sense to make this a `property`?

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        # type: () -> List[Device]
+        return self._data
+
+    def wal_devices(self):
+        # type: () -> List[Device]
+        return self._wal
+
+    def db_devices(self):
```
Comment (votdev) [inline src/python-common/ceph/deployment/drive_selection/selector.py]: Wouldn't it make sense to make this a `property`?

File: src/python-common/ceph/deployment/drive_selection/selector.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        # type: () -> List[Device]
+        return self._db
+
+    def journal_devices(self):
```
Comment (votdev) [inline src/python-common/ceph/deployment/drive_selection/selector.py]: Wouldn't it make sense to make this a `property`?

File: src/python-common/ceph/tests/test_disk_selector.py (hunk 9 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/tests/test_disk_selector.py]: Right, the way EqualityMatcher os implemented, this could be any string.

File: src/python-common/ceph/tests/test_disk_selector.py (hunk 10 above)
Comment (sebastian-philipp) [reply src/python-common/ceph/tests/test_disk_selector.py]: Right, the way EqualityMatcher os implemented, this could be any string.

**Comment by sebastian-philipp** [comment]
@votdev would you mind if we merge this, despite the nits?

**Comment by votdev** [comment]
> @votdev would you mind if we merge this, despite the nits?

Yes, no problem. Maybe you can fix it in a follow up PR.
